### PR TITLE
Non-multipart file upload

### PR DESCRIPTION
**Pull request by zachmullen**
This mode has a couple significant advantages:

1. It's much faster -- the cherrypy server does not need to
   expend resources searching the body for multipart chunk
   delimiters.
2. The cherrypy server does not write the file part to disk,
   and the request handler is invoked immediately rather than
   waiting for the whole request body. This removes the
   noticeable pause during uploading when the chunk was being
   processed.

@Xarthisius this fixes the issue you noticed with uploads being slow.

### CODE REVIEW

**Comment by Xarthisius** [comment]
@zachmullen This is awesome! With those changes upload speed I get is comparable with raw `scp`

**Comment by zachmullen** [comment]
This is now ready for review.

**Comment by manthey** [comment]
I realize that non-multipart uploads are faster, but is there a reason to deprecate the multipart uploads, rather than just accept both in general?

**Comment by zachmullen** [comment]
Mostly just since they are slower and put more load on the server. I anticipate they'll be around for a couple years yet, but I wanted to give us the option in the far future to remove it if we want.

**Comment by mgrauer** [comment]
Does deprecated mean something specific?  Like we'll support it until the next major version or for 6 months?  I'd like to work towards getting it to mean something precise and documenting that.  This may already exist.

**Comment by zachmullen** [comment]
I was avoiding doing that, but it would be a good thing to do. A lot of tools will leave something deprecated for a full major release, but our major release cycles are really long (at least so far). Maybe we should base it on wall time + major release...?

**Comment by zachmullen** [comment]
It occurs to me that as written, this will make it so newer versions of girder-client would not be able to upload to older versions of girder server. I think I'll add a check for server version and do multipart if the server version is < 2.2. Does that make sense?

**Comment by jcfr** [comment]
I tried to check if caching of version worked running the following code in a python console:

```
>>> from  girder_client import GirderClient
>>> gc = GirderClient(apiUrl="http://localhost:8080/api/v1")
>>> gc.authenticate(apiKey="<MY_TOKEN>")
>>> gc.getServerVersion()
['2', '1', '1']
```
Looking at the log confirms it worked.


That said, it also indicated that the version currently associated with my girder build is `2.1.1`. 

Since the girder client test for the version of Girder to choose the upload method (using ``self.getServerVersion() < ['2', '2']``), I wasn't able to validate the new approach.


Nevertheless, I tried to use the updated client and it confirmed the old approach is still supported by both:
1. Upload data on https://data.kitware.com which is older than Girder 2.2 and it worked
2. Upload data on http://localhost:8080 and it worked. The log below confirm it:

```
127.0.0.1 - - [09/Mar/2017:14:20:10] "GET /api/v1/item/58c1aaea4154cd1bd7d748db/files HTTP/1.1" 200 2 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:14:20:10] "POST /api/v1/file?name=BrainAtlas2012.mrb&size=174251653&parentId=58c1aaea4154cd1bd7d748db&parentType=item HTTP/1.1" 200 897 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:14:20:10] "GET /api/v1/system/version HTTP/1.1" 200 201 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:14:20:12] "POST /api/v1/file/chunk?offset=0&uploadId=58c1aaea4154cd1bd7d748dc HTTP/1.1" 200 904 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:14:20:15] "POST /api/v1/file/chunk?offset=67108864&uploadId=58c1aaea4154cd1bd7d748dc HTTP/1.1" 200 905 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:14:20:16] "POST /api/v1/file/chunk?offset=134217728&uploadId=58c1aaea4154cd1bd7d748dc HTTP/1.1" 200 333 "" "python-requests/2.10.0"
```

Next step of the review process:

* discuss how to improve the feature "does girder support non-multipart file upload" detection to not rely on girder version. I am thinking of adding a method to the client that would checking if the API call has the expected signature.

* test upload with Girder <=2.2 and >2.2 when the assetstore is amazon s3.

**Comment by zachmullen** [comment]
Just for info, I was planning to cut version 2.2 shortly after this and #1838 go in, and also cut a release of the python client. That way, even those tracking master will get these new changes in relatively short order.

**Comment by jcfr** [comment]
With the improved detection of the nonmultipart feature support, I confirm that the proper api endpoint is used:

```
127.0.0.1 - - [09/Mar/2017:20:33:46] "POST /api/v1/api_key/token?key=<MY_TOKEN> HTTP/1.1" 200 213 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:20:33:46] "POST /api/v1/item?reuseExisting=False&description=&name=BrainAtlas2012.mrb&folderId=574305a44154cd4e67c1d836 HTTP/1.1" 200 362 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:20:33:46] "GET /api/v1/item/58c2027a4154cd1bd7e245ee/files HTTP/1.1" 200 2 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:20:33:47] "POST /api/v1/file?size=174251653&name=BrainAtlas2012.mrb&parentId=58c2027a4154cd1bd7e245ee&parentType=item HTTP/1.1" 200 897 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:20:33:47] "GET /api/v1/describe HTTP/1.1" 200 103755 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:20:33:47] "POST /api/v1/file/chunk?offset=0&uploadId=58c2027b4154cd1bd7e245ef HTTP/1.1" 200 904 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:20:33:47] "POST /api/v1/file/chunk?offset=67108864&uploadId=58c2027b4154cd1bd7e245ef HTTP/1.1" 200 905 "" "python-requests/2.10.0"
127.0.0.1 - - [09/Mar/2017:20:33:47] "POST /api/v1/file/chunk?offset=134217728&uploadId=58c2027b4154cd1bd7e245ef HTTP/1.1" 200 333 "" "python-requests/2.10.0"

```

**Comment by jcfr** [comment]
cc: @mgrauer

**Comment by jcfr** [comment]
Note that I will fix the docstring warning soon:

```
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:9: ERROR: Unexpected indentation.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:13: ERROR: Unexpected indentation.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:15: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:17: ERROR: Unexpected indentation.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:22: ERROR: Unexpected indentation.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:28: WARNING: Definition list ends without a blank line; unexpected unindent.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:29: SEVERE: Unexpected section title.

},
...
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:30: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:31: WARNING: Definition list ends without a blank line; unexpected unindent.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:32: SEVERE: Unexpected section title.

}.
...
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:33: WARNING: Definition list ends without a blank line; unexpected unindent.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:34: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py:docstring of girder_client.GirderClient.getServerAPIDescription:36: WARNING: Block quote ends without a blank line; unexpected unindent.

```

**Comment by zachmullen** [comment]
@jcfr I don't think this is a safe enough way to detect the support of the new upload mode -- the only reason "chunk" was removed from the params is because our autoDescribeRoute is insufficiently powerful to handle a param type that could be either in the body or in the params, but in the future if we change that, or remove support for multipart upload, it would break your check.

There are only two ways I'm truly comfortable with their level of future-proofing: 1. check the server version as I originally did or 2. explicitly add something to the API that signals the presence of the new feature, but is not tied to any part of the swagger description that might change in the future.

**Comment by jcfr** [comment]
Using the client (with non-multipart enabled) to upload files on a girder instance with s3 asset store fails with the following:

```
  File "/home/jcfr/Projects/girder/clients/python/girder_client/__init__.py", line 395, in sendRestRequest
    status=result.status_code, url=result.url, method=method, text=result.text)
girder_client.HttpError: HTTP error 500: POST http://localhost:8080/api/v1/file/chunk?offset=0&uploadId=58c705524154cd6574f7a692
Response text: {"message": "AttributeError: AttributeError(\"'RequestBodyStream' object has no attribute 'tell'\",)", "trace": [
  "<FrameSummary file /home/jcfr/Projects/girder/girder/api/rest.py, line 561 in endpointDecorator>", 
  "<FrameSummary file /home/jcfr/Projects/girder/girder/api/rest.py, line 1123 in POST>", 
  "<FrameSummary file /home/jcfr/Projects/girder/girder/api/rest.py, line 884 in handleRoute>", 
  "<FrameSummary file /home/jcfr/Projects/girder/girder/api/access.py, line 77 in wrapped>", 
  "<FrameSummary file /home/jcfr/Projects/girder/girder/api/describe.py, line 575 in wrapped>", 
  "<FrameSummary file /home/jcfr/Projects/girder/girder/api/v1/file.py, line 222 in readChunk>", 
  "<FrameSummary file /home/jcfr/Projects/girder/girder/models/upload.py, line 143 in handleChunk>", 
  "<FrameSummary file /home/jcfr/Projects/girder/girder/utility/s3_assetstore_adapter.py, line 248 in uploadChunk>", 
  "<FrameSummary file /home/jcfr/Projects/girder/girder/utility/s3_assetstore_adapter.py, line 329 in _proxiedUploadChunk>", 
  "<FrameSummary file /home/jcfr/.virtualenvs/girder_env/lib/python3.5/site-packages/boto/s3/key.py,
 line 1224 in set_contents_from_file>"], "type": "internal"}

```

It works if I use the slower multipart approach.

**Comment by jcfr** [comment]
After adding:

```
    def seek(self, *args, **kwargs):
        pass

    def tell(self, *args, **kwargs):
        return 0
```

to `girder.utility.RequestBodyStream`.

I get:

```
girder_client.HttpError: HTTP error 500: POST http://localhost:8080/api/v1/file/chunk?offset=0&uploadId=58c7080f4154cd65741cadf8
Response text: {
  "message": "S3ResponseError: S3ResponseError: 400 Bad Request\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n
<Error>
<Code>BadDigest</Code>
<Message>The Content-MD5 you specified did not match what we received.</Message>
<ExpectedDigest>03028c01600bbd34c230e5e5b1228bf0</ExpectedDigest>
<CalculatedDigest>1B2M2Y8AsgTpgAmY7PhCfg==</CalculatedDigest>
<RequestId>6DEA428422005A4C</RequestId>
<HostId>zSbLJFuVZnr8kU5VkwojIU9wPmMFI/jCfn1PeC9yOo9eJFfqBUjjCko+DFCEUqBJO3mp+RfWTGI=</HostId>
</Error>", 
"trace": ["<FrameSummary file /home/jcfr/Projects/girder/girder/api/rest.py, line 561 in endpointDecorator>", 
"<FrameSummary file /home/jcfr/Projects/girder/girder/api/rest.py, line 1123 in POST>", 
"<FrameSummary file /home/jcfr/Projects/girder/girder/api/rest.py, line 884 in handleRoute>", 
"<FrameSummary file /home/jcfr/Projects/girder/girder/api/access.py, line 77 in wrapped>", 
"<FrameSummary file /home/jcfr/Projects/girder/girder/api/describe.py, line 575 in wrapped>", 
"<FrameSummary file /home/jcfr/Projects/girder/girder/api/v1/file.py, line 222 in readChunk>", 
"<FrameSummary file /home/jcfr/Projects/girder/girder/models/upload.py, line 143 in handleChunk>", 
"<FrameSummary file /home/jcfr/Projects/girder/girder/utility/s3_assetstore_adapter.py, line 248 in uploadChunk>", 
"<FrameSummary file /home/jcfr/Projects/girder/girder/utility/s3_assetstore_adapter.py, line 329 in _proxiedUploadChunk>",
"<FrameSummary file /home/jcfr/.virtualenvs/girder_env/lib/python3.5/site-packages/boto/s3/key.py, line 1293 in set_contents_from_file>", 
"<FrameSummary file /home/jcfr/.virtualenvs/girder_env/lib/python3.5/site-packages/boto/s3/key.py, line 750 in send_file>", 
"<FrameSummary file /home/jcfr/.virtualenvs/girder_env/lib/python3.5/site-packages/boto/s3/key.py, line 951 in _send_file_internal>", 
"<FrameSummary file /home/jcfr/.virtualenvs/girder_env/lib/python3.5/site-packages/boto/s3/connection.py, line 668 in make_request>",
 "<FrameSummary file /home/jcfr/.virtualenvs/girder_env/lib/python3.5/site-packages/boto/connection.py, line 1071 in make_request>", 
"<FrameSummary file /home/jcfr/.virtualenvs/girder_env/lib/python3.5/site-packages/boto/connection.py, line 940 in _mexe>", 
"<FrameSummary file /home/jcfr/.virtualenvs/girder_env/lib/python3.5/site-packages/boto/s3/key.py, line 884 in sender>"
```

**Comment by zachmullen** [comment]
This is now ready for a final review pass.

**Comment by jcfr** [comment]
For sake of comparison, for 622M, time to transfer:
* using multipart upload to S3  ``3m2.422s``
* using non-multipart upload to S3  ``2m52.405s``

File: tests/cases/py_client/lib_test.py [hunk 1]
```diff
@@ -316,6 +316,44 @@ def progressCallback(info):
...
             self.assertEqual(file['mimeType'], 'text/plain')
 
+    def testUploadNonMultipartVersionGreaterOrEqual22(self):
+        for version in ["2.2.0", "2.2.1", "2.3", "3.0", "3.1"]:
+            with mock.patch.object(
+                    self.client, 'getServerVersion', return_value=version.split(".")):
+                self._testUploadMethod(expected_non_multipart_hits=1, expected_multipart_hits=0)
+
+    def testUploadMultipartVersionLess22(self):
+        for version in ["1.4.1", "1.6.0", "2.1.0", "2.1.9"]:
```
Comment (zachmullen) [inline tests/cases/py_client/lib_test.py]: Minor thing, but let's use single quotes everywhere for consistency

**Comment by zachmullen** [comment]
@jcfr thanks for the nice tests! 👍 

This is good to go from my perspective, if you are happy with it, please mark as approved and we can merge. 🎉

**Pull request merged**